src/google/protobuf/util/field_mask_util.cc

```diff
@@ -469,8 +469,13 @@ void FieldMaskTree::MergeMessage(const Node* node, const Message& source,
                    << "have sub-fields.";
         continue;
       }
-      MergeMessage(child, source_reflection->GetMessage(source, field), options,
-                   destination_reflection->MutableMessage(destination, field));
+
+      if (!source_reflection->HasField(source, field)) {
+        destination_reflection->ClearField(destination, field);
```

Comment:
Actually in the case where the source message does not have the given field, I think the right thing to do is just leave the destination field alone instead of clearing it.

Reply:
The main reason for doing this is described on example in description for this PR. 

If shortly, it just copies the behaviour for primitives which are not specified in the FieldMask, since all of them are cleared in the if statement below. 

```cpp
if (!field->is_repeated()) {
       switch (field->cpp_type()) {
 #define COPY_VALUE(TYPE, Name)                                              \
   case FieldDescriptor::CPPTYPE_##TYPE: {                                   \
     if (source_reflection->HasField(source, field)) {                       \
       destination_reflection->Set##Name(                                    \
           destination, field, source_reflection->Get##Name(source, field)); \
     } else {                                                                \
       destination_reflection->ClearField(destination, field);               \
     }                                                                       \
     break;                                                                  \
   }
   ...
   ...
   ...
}
```

Reply:
OK, I think you are right. That does make sense that the behavior should be consistent with how it works for primitive types. I can see how one might want to use a field mask to clear a particular field. Is this covered by any tests, though? Maybe I am missing something but it looks like the new test doesn't cover the case of using a field mask to clear a field that already exists in the destination message.

Reply:
I believe I understood what you meant, but just want to double check it first :)

Am I right that you are looking for the test, which verifies the following:

We are having two messages like that 
```proto
message M {
    string first = 1;
    MM second = 2;
}

message MM {
     int64 value = 1;
}
```

* We have a source message `M` which **does not** have field *second*. 

* The field mask will be definitely having `second.value` as one of the paths.

* Destination message `M` which **has** something stored in field *second*.

As a result, after calling `MergeMessage` we should expect that destination will not have anything stored in field *second*.

Could you confirm, that this is the scenario we would like to cover?

Reply:
Right, that is exactly the scenario I was thinking of.

But actually, now that I have thought about it some more, I wonder if the logic still needs to change somewhat. Let's say the field mask is "a.b.c.d.e.f" and the source message does not have field "a". In that case the destination proto will have "a" completely cleared, even though the field mask is really just indicating a particular deeply nested field within "a". So I think the right answer is that we shouldn't clear any field until we get down to the very last element of the field mask path.